### PR TITLE
feat: Some lemmas about `abs`, `sqrt` , and `ZMod`

### PR DESCRIPTION
**Pull request by javra**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/GeomSum.lean [hunk 1]
```diff
@@ -587,6 +587,17 @@ theorem geom_sum_neg_iff [LinearOrderedRing α] (hn : n ≠ 0) :
 
 end Order
 
+theorem pow_sub_pow_le [StrictOrderedCommRing α]
```
Comment (YaelDillies) [inline Mathlib/Algebra/GeomSum.lean:590]: This should be enough
```suggestion
theorem pow_sub_pow_le [OrderedCommRing α]
```

File: Mathlib/Data/Option/Basic.lean [hunk 2]
```diff
@@ -438,4 +438,10 @@ theorem elim_comp₂ (h : α → β → γ) {f : γ → α} {x : α} {g : γ →
 theorem elim_apply {f : γ → α → β} {x : α → β} {i : Option γ} {y : α} :
     i.elim x f y = i.elim (x y) fun j => f j y := by rw [elim_comp fun f : α → β => f y]
 
+@[simp]
+theorem get!_some [Inhabited α] (a : α) : (Option.some a).get! = a := by rfl
```
Comment (YaelDillies) [inline Mathlib/Data/Option/Basic.lean:442]: Is this not simply `rw [Option.get!]`?

File: Mathlib/Data/ZMod/Basic.lean [hunk 3]
```diff
@@ -934,6 +934,19 @@ theorem val_eq_zero : ∀ {n : ℕ} (a : ZMod n), a.val = 0 ↔ a = 0
...
+    a.val = 1 ↔ a = 1 := by
+  rcases n with (⟨⟩|⟨n⟩)
+  · have := h.prop
+    simp at this
+  · rcases n with (⟨⟩|⟨n⟩)
+    · have := h.prop
+      norm_num at this
+    · rw [Fin.ext_iff]
+      simp only [Fin.val_one]
+      exact Iff.rfl
```
Comment (YaelDillies) [inline Mathlib/Data/ZMod/Basic.lean:948]: See https://github.com/YaelDillies/LeanAPAP/blob/master/LeanAPAP/Mathlib/Data/ZMod/Basic.lean#L25-L30 for ideas

File: Mathlib/Data/Option/Basic.lean (hunk 2 above)
Comment (javra) [reply Mathlib/Data/Option/Basic.lean:442]: No, it's the same as `rw [Option.get!]; dsimp only` in order to reduce the `match` statement

File: Mathlib/Algebra/GeomSum.lean [hunk 4]
```diff
@@ -587,6 +587,17 @@ theorem geom_sum_neg_iff [LinearOrderedRing α] (hn : n ≠ 0) :
 
 end Order
 
+theorem pow_sub_pow_le [OrderedCommRing α]
+    {a b : α} (n : ℕ) (hb : 0 ≤ b) (ha : b ≤ a) :
```
Comment (YaelDillies) [inline Mathlib/Algebra/GeomSum.lean:591]: The convention is rather
```suggestion
    {a b : α} (n : ℕ) (hb : 0 ≤ b) (hba : b ≤ a) :
```

File: Mathlib/Algebra/GeomSum.lean [hunk 5]
```diff
@@ -587,6 +587,17 @@ theorem geom_sum_neg_iff [LinearOrderedRing α] (hn : n ≠ 0) :
...
+    a ^ n - b ^ n ≤ n * a ^ (n - 1) * (a - b) := by
+  rw [← geom_sum₂_mul]
+  apply mul_le_mul_of_nonneg_right _ (sub_nonneg.mpr ha)
+  rw [← @geom_sum₂_self]
+  apply Finset.sum_le_sum
+  intros
+  apply mul_le_mul_of_nonneg_left _ (pow_nonneg (le_trans hb ha) _)
+  apply pow_le_pow_left hb ha
```
Comment (YaelDillies) [inline Mathlib/Algebra/GeomSum.lean:599]:
```suggestion
  gcongr
  -- Maybe something here
```

File: Mathlib/Algebra/GeomSum.lean [hunk 6]
```diff
@@ -587,6 +587,17 @@ theorem geom_sum_neg_iff [LinearOrderedRing α] (hn : n ≠ 0) :
 
 end Order
 
+theorem pow_sub_pow_le [OrderedCommRing α]
+    {a b : α} (n : ℕ) (hb : 0 ≤ b) (ha : b ≤ a) :
+    a ^ n - b ^ n ≤ n * a ^ (n - 1) * (a - b) := by
+  rw [← geom_sum₂_mul]
+  apply mul_le_mul_of_nonneg_right _ (sub_nonneg.mpr ha)
+  rw [← @geom_sum₂_self]
```
Comment (YaelDillies) [inline Mathlib/Algebra/GeomSum.lean:595]: You could try using `gcongr` even earlier, actually

File: Mathlib/Algebra/Order/Group/Abs.lean [hunk 7]
```diff
@@ -130,6 +130,15 @@ variable [CovariantClass α α (swap (· * ·)) (· ≤ ·)]
...
 #align lattice_ordered_comm_group.abs_abs abs_abs
 
+-- should those be simp lemmas?
+@[to_additive]
+theorem abs_ite_inv (x : α) (p : Prop) [Decidable p] : |if p then x⁻¹ else x|ₘ = |x|ₘ := by
+  split_ifs <;> simp
+
+@[to_additive]
+lemma abs_ite_inv'  (x : α) (p : Prop) [Decidable p] : |if p then x else x⁻¹|ₘ = |x|ₘ := by
+  split_ifs <;> simp
```
Comment (YaelDillies) [inline Mathlib/Algebra/Order/Group/Abs.lean:140]: I think they will hardly ever apply, but conversely it doesn't hurt tagging them simp

File: Mathlib/Data/Nat/Cast/Field.lean [hunk 8]
```diff
@@ -42,6 +42,21 @@ theorem cast_div_div_div_cancel_right [DivisionSemiring α] [CharZero α] {m n d
...
+theorem cast_div_eq_sub [DivisionRing α] [CharZero α] {a b : ℕ} :
+    ↑(a / b) = (a : α) / (b : α) - ↑(a % b) / (b : α) := by
+  by_cases hb : b = 0
+  · subst hb; simp
+  · rw [Nat.div_eq_sub_mod_div,
+      Nat.cast_div (Nat.dvd_sub_mod a) (Nat.cast_ne_zero.mpr hb),
+      Nat.cast_sub (Nat.mod_le a b), sub_div]
+
+theorem cast_gt [LinearOrderedField α] [ZeroLEOneClass α] {a b : ℕ} (hb : 0 < b):
+    ((a / b : ℕ) : α) > (a : α) / (b : α) - 1 := by
```
Comment (YaelDillies) [inline Mathlib/Data/Nat/Cast/Field.lean:54]: Please never use `>`. Always prefer `<` over it. Also you're using way too many type ascriptions.
```suggestion
    (a / b : α) - 1 < ↑(a / b) := by
```

File: Mathlib/Data/Nat/Cast/Field.lean [hunk 9]
```diff
@@ -42,6 +42,21 @@ theorem cast_div_div_div_cancel_right [DivisionSemiring α] [CharZero α] {m n d
...
 
+theorem cast_div_eq_sub [DivisionRing α] [CharZero α] {a b : ℕ} :
+    ↑(a / b) = (a : α) / (b : α) - ↑(a % b) / (b : α) := by
+  by_cases hb : b = 0
+  · subst hb; simp
+  · rw [Nat.div_eq_sub_mod_div,
+      Nat.cast_div (Nat.dvd_sub_mod a) (Nat.cast_ne_zero.mpr hb),
+      Nat.cast_sub (Nat.mod_le a b), sub_div]
+
+theorem cast_gt [LinearOrderedField α] [ZeroLEOneClass α] {a b : ℕ} (hb : 0 < b):
```
Comment (YaelDillies) [inline Mathlib/Data/Nat/Cast/Field.lean:53]: This name is too generic.
```suggestion
theorem lt_cast_div [LinearOrderedField α] [ZeroLEOneClass α] {a b : ℕ} (hb : 0 < b) :
```
or even
```suggestion
theorem div_sub_one_lt_cast_div [LinearOrderedField α] [ZeroLEOneClass α] {a b : ℕ} (hb : 0 < b) :
```
for discoverability

**Review by YaelDillies — COMMENTED** [review]
It would be easier to review (mostly, you wouldn't wait for so long) if you opened one PR per file instead of putting all the changes together. Please prepend `feat: ` to each PR title to indicate that you're adding new mathematical content. You can reuse the current PR to be about one of the files.

**Comment by javra** [comment]
@YaelDillies Sorry, I'm in no rush to get them into mathlib, but I'm short on time, this is not my main job.

**Comment by javra** [comment]
Not sure about the placement of `pow_sub_pow_le` in Algebra.GeomSum, the proof *uses* the geometric sum, but it doesn't appear in the actual statement..

**Comment by loefflerd** [comment]
> _Yael_: It would be easier to review (mostly, you wouldn't wait for so long) if you opened one PR per file instead of putting all the changes together.
[...]
> _Jakob_: Sorry, I'm in no rush to get them into mathlib, but I'm short on time, this is not my main job.

Jakob: are you aware that the mathlib maintainer / reviewer team all have other jobs as well? Unless you're willing to meet reviewers half-way and restructure your PRs a little to make them easier to review, they'll likely stay as "awaiting review" forever. This is particularly the case if you make comments like this one, which (whether intentionally or not) creates the impression that you think your time is more valuable than the reviewers'. 

To me this PR looks like a rag-bag of random lemmas, all completely undocumented, with no clear unifying theme and mostly without any obvious applications. Why does mathlib need these lemmas? A charitable interpretation is that they are the prerequisites for some more substantial code that you are going to add in a future PR; but then you should probably say so.

**Comment by javra** [comment]
Sorry, that was not at all the intention oh my above comment. This PR was made as a response to [this thread](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/bunch.20of.20lemmas.20to.20upstream/near/437405169). All of those lemmas are in use in other projects than mathlib, but I thought they're better placed in mathlib itself.

**Comment by javra** [comment]
Will close this for now, not too happy that I didn't get a reply to that question in the Zulip thread of "is it worth PRing lemmas like these". Would it be enough separation to split them into one PR concerning `abs` and one concerning `ZMod`?

**Comment by loefflerd** [comment]
> Will close this for now, not too happy that I didn't get a reply to that question in the Zulip thread of "is it worth PRing lemmas like these"

It wasn't my intention to suggest that the lemmas are not "worth PRing". Indeed, Yael and I are trying to encourage you to make more PR's, not less. 

> Would it be enough separation to split them into one PR concerning abs and one concerning ZMod?

It would, if all the lemmas genuinely concerned one or other of those things. But this is not the case, e.g. the two `Option.get!` lemmas don't seem to have anything to do with either `abs` or `ZMod`.

**Pull request closed without merging**